jooqdb/src/main/java/org/cristalise/storage/jooqdb/JooqClusterStorage.java

```diff
@@ -1,339 +1,426 @@
+    protected List<JooqDomainHandler>           domainHandlers = new ArrayList<JooqDomainHandler>();
+    protected ConcurrentHashMap<Object, DSLContext> contextMap = new ConcurrentHashMap<Object, DSLContext>(); 
+
+    @Override
+    public void open(Authenticator auth) throws PersistencyException {
+        initialiseHandlers();
+    }
+
+    /**
+     * Initialise internal handlers for all ClusterTypes and all the DomainHandlers-
+     *
+     * @throws PersistencyException Error during initialise ...
+     */
+    public void initialiseHandlers() throws PersistencyException {
+        Logger.msg(1, "JooqClusterStorage.initialiseHandlers() - Starting with standard hadlers.");
+
+        jooqHandlers.put(ClusterType.PROPERTY,   new JooqItemPropertyHandler());
+        jooqHandlers.put(ClusterType.OUTCOME,    new JooqOutcomeHandler());
+        jooqHandlers.put(ClusterType.VIEWPOINT,  new JooqViewpointHandler());
+        jooqHandlers.put(ClusterType.LIFECYCLE,  new JooqLifecycleHandler());
+        jooqHandlers.put(ClusterType.COLLECTION, new JooqCollectionHadler());
+        jooqHandlers.put(ClusterType.HISTORY,    new JooqHistoryHandler());
+        jooqHandlers.put(ClusterType.JOB,        new JooqJobHandler());
+        jooqHandlers.put(ClusterType.ATTACHMENT, new JooqOutcomeAttachmentHandler());
+
+        DSLContext context = JooqHandler.connect();
+
+        context.transaction(nested -> {
+            for (JooqHandler handler: jooqHandlers.values()) handler.createTables(DSL.using(nested));
+        });
+
+        try {
+            String handlers = Gateway.getProperties().getString(JOOQ_DOMAIN_HANDLERS, "");
+
+            for(String handlerClass: StringUtils.split(handlers, ",")) {
+                if (!handlerClass.contains(".")) handlerClass = "org.cristalise.storage."+handlerClass;
+
+                Logger.msg(1, "JooqClusterStorage.initialiseHandlers() - Instantiate domain handler:"+handlerClass);
+
+                domainHandlers.add( (JooqDomainHandler) Class.forName(handlerClass).newInstance());
+            }
+        }
+        catch (InstantiationException | IllegalAccessException | ClassNotFoundException ex) {
+            Logger.error("JooqClusterStorage could not instantiate domain handler");
+            Logger.error(ex);
+            throw new PersistencyException("JooqClusterStorage could not instantiate domain handler:"+ex.getMessage());
+        }
+
+        if (! Gateway.getProperties().getBoolean(JOOQ_DISABLE_DOMAIN_CREATE, false)) {
+            context.transaction(nested -> {
+                for (JooqDomainHandler handler: domainHandlers) handler.createTables(DSL.using(nested));
+            });
+        }
+    }
+
+    public void dropHandlers() throws PersistencyException {
+        DSLContext context = JooqHandler.connect();
+        for (JooqHandler handler: jooqHandlers.values()) handler.dropTables(context);

```

Comment:
Shouldn't this be also in a `context.transaction(nested -> {` block?

Reply:
Changes will be on next commit.